### DISCUSSION TITLE
About using NMNIST dataset, num_steps=300, dt=1000

### DISCUSSION THREAD

**Discussion by mountains-high** [jeshraghian/snntorch#57] [Q&A] [answered]
Good day!

I wanted to know what does mean num_steps = 300. Does it mean that to use only 300 events?
Like we have 60000 samples in NMNIST, and each sample has around 4500 events. But if we set dt =1000,
and num_steps = 300, so we use only 300 events out of 4500?

Thank you!

**Comment by jeshraghian** [comment, marked as answer]
Hi @asselkembay! 

For future reference, please use the Discussions forum for questions so that the Issues can be used for tracking bugs and adding features :)

`num_steps` refers to the number of time steps being simulated, and `dt` is the integration window of the events in microseconds.

In a recurrent neural network, the number of steps corresponds to the length of the input sequence. You can change the sequence length of the input using `num_steps`. 

The NMNIST dataset was filmed using a DVS Camera which generates a variable number of events at any time step.
If `dt=1`, then your input would catch very few events at each time step. 
If `dt=1000`, then you are passing in many more (1ms worth) events at each time step. 

As you increase `dt`, the faster the NMNIST will be passed into the neural network.
This means that for a fixed `num_steps`, the final steps will have no more events. 
The default values are widely used amongst papers.

I recommend opening up this notebook, playing with the values, and visualizing them to see how they are related:
https://github.com/jeshraghian/snntorch/blob/master/examples/NMNIST.ipynb

**Reply by mountains-high** [reply]
Good day, 
Thank you for your consideration. 
I’m sorry I didn’t notice that. Yeap, in the future I will be using Discussion. 

If the number of **timestamps** are equal to around 4500, then it means that we have 4500 events for each one of a sample (we have 60000 samples) because so many times events were recorded. 

Let’s say event_0 is filmed or recorded, captured at time 1415us, and event_1 is captured at 1530us. 
```According to dt = 1000 and num_steps = 300```
Does it mean that dt = 1000us(dt = 1ms), so that the first event_0 assumed to be recorded at T = 0ms, then, event_1 = 1ms, event_2 = 2ms, … event_299 = 300ms?

So that you used only 300 events with the range of 300ms with time step 1ms. 
Or am I misunderstanding something?

Thank you very much!

**Reply by jeshraghian** [reply]
> If the number of timestamps are equal to around 4500, then it means that we have 4500 events for each one of a sample (we have 60000 samples) because so many times events were recorded.

-So this is incorrect. If you record for 4500 'timestamps', then it is possible to have way more events, or way less events. Each time stamp may have multiple events, or it may have zero events. 

I'll use your example to illustrate.
Say event_0 is recorded at t=200us, event_1=803us.
If `dt=1000` (1ms), then T=1ms will have both event_0 and event_1 appear at the same time. 
Then event_2=1400us will appear at T=2ms. 

As a rule of thumb, say you change `dt` from 1000 to 500.
Then you should change `num_steps` from 300 to 600 to capture the full duration.

If you change `dt` from 1000 to 2000.
Then you should change `num_steps` from 300 to 150. Otherwise, the final 150 steps will have no events.

Hope this makes it clear! I highly recommend opening the notebook sent above with Google Colab. 
That should help make it more clear if you run the animations.

**Reply by mountains-high** [reply]
Thank you Mr. @jeshraghian for your detailed explanation.
